File: Estructurales Java/Decorator.java
// Decorator Pattern
abstract class Component {
    public abstract String operation();
}

class ConcreteComponent extends Component {
    @Override
    public String operation() {
        return "ConcreteComponent";
    }
}

abstract class Decorator extends Component {
    protected Component component;
    
    public Decorator(Component component) {
        this.component = component;
    }
    
    @Override
    public String operation() {
        return component.operation();
    }
}

class ConcreteDecoratorA extends Decorator {
    public ConcreteDecoratorA(Component component) {
        super(component);
    }
    
    @Override
    public String operation() {
        return "ConcreteDecoratorA(" + component.operation() + ")";
    }
}

class ConcreteDecoratorB extends Decorator {
    public ConcreteDecoratorB(Component component) {
        super(component);
    }
    
    @Override
    public String operation() {
        return "ConcreteDecoratorB(" + component.operation() + ")";
    }
}

// Usage
class DecoratorDemo {
    public static void main(String[] args) {
        Component simple = new ConcreteComponent();
        System.out.println("Client: I got a simple component: " + simple.operation());
        
        Component decorator1 = new ConcreteDecoratorA(simple);
        Component decorator2 = new ConcreteDecoratorB(decorator1);
        System.out.println("Client: Now I've got a decorated component: " + decorator2.operation());
    }
}

File: Estructurales Python/decorator.py
# Decorator Pattern
from abc import ABC, abstractmethod

class Component(ABC):
    @abstractmethod
    def operation(self):
        pass

class ConcreteComponent(Component):
    def operation(self):
        return "ConcreteComponent"

class Decorator(Component):
    def __init__(self, component: Component):
        self._component = component
    
    def operation(self):
        return self._component.operation()

class ConcreteDecoratorA(Decorator):
    def operation(self):
        return f"ConcreteDecoratorA({self._component.operation()})"

class ConcreteDecoratorB(Decorator):
    def operation(self):
        return f"ConcreteDecoratorB({self._component.operation()})"

# Usage
if __name__ == "__main__":
    simple = ConcreteComponent()
    print(f"Client: I got a simple component: {simple.operation()}")
    
    decorator1 = ConcreteDecoratorA(simple)
    decorator2 = ConcreteDecoratorB(decorator1)
    print(f"Client: Now I've got a decorated component: {decorator2.operation()}")
